core/modules/cart/store/actions.ts

```diff
@@ -572,12 +587,14 @@ const actions: ActionTree<CartState, RootState> = {
     }
 
     /** helper - sub method to react for the server response after the sync */
-    const _afterServerItemUpdated = async function ({ dispatch, commit }, event, clientItem = null) {
+    const _afterServerItemUpdated = async function ({ dispatch, commit }, event, clientItem = null, serverItem = null) {
       Logger.debug('Cart item server sync' + event, 'cart')()
       diffLog.serverResponses.push({ 'status': event.resultCode, 'sku': clientItem.sku, 'result': event })
       if (event.resultCode !== 200) {
         // TODO: add the strategy to configure behaviour if the product is (confirmed) out of the stock
-        if (clientItem.server_item_id) {
+        if (!serverItem) {
```

Comment:
I’m not sure why this line has been added? There is another place where removeItem is executed

Reply:
I mean line 610: when there is no serverItem, calling the “removeItem” which executed the recursive “sync” + “merge” is pointless and could cause some side effects. We should just remove the item client side as it’s done in 610 I belive. Please let me know if there is some other reason why it was done differently?

Reply:
Yeah, the reason why I added this line it's because it doesn't deleting the item - it's the case when product is "out of stock".

Considering following steps:
1. Add item to the cart (size:S, color: black)
2. Change size to XS (that size is out of stock)
3. You receive the notification from the server "out of stock" but the product is still in the cart

I've noticed, it wasn't deleted because we have _server_item_id_ set (mutation: CART_DEL_NON_CONFIRMED_ITEM)

Reply:
That's a good point! OK, so `removeItem` should be fine (as it won't sync when there is `server_item_id` set).